oximeter/impl/src/schema/ir.rs

```diff
@@ -0,0 +1,1288 @@
+                        {
+                            let field_schema = construct_field_schema(
+                                &self.fields,
+                                target_name,
+                                target_fields,
+                                metric_name,
+                                &metric_field_names,
+                            )?;
+                            let authz_scope = extract_authz_scope(
+                                metric_name,
+                                self.target.authz_scope,
+                                &field_schema,
+                            )?;
+                            let timeseries_name = TimeseriesName::try_from(
+                                format!("{}:{}", target_name, metric_name),
+                            )?;
+                            let version = NonZeroU8::new(*version).unwrap();
+                            let description = TimeseriesDescription {
+                                target: self.target.description.clone(),
+                                metric: metric.description.clone(),
+                            };
+                            let schema = TimeseriesSchema {
+                                timeseries_name: timeseries_name.clone(),
+                                description,
+                                field_schema,
+                                datum_type: metric.datum_type,
+                                version,
+                                authz_scope,
+                                units: metric.units,
+                                created: Utc::now(),
+                            };
+                            if let Some(old) = timeseries
+                                .insert((timeseries_name, version), schema)
+                            {
+                                return defn_error(format!(
+                                    "Timeseries '{}' version {} is duplicated",
+                                    old.timeseries_name, old.version,
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(timeseries.into_values().collect())
+    }
+}
+
+#[derive(Clone, Debug)]
+enum CurrentVersion {
+    Active { version: NonZeroU8, fields: BTreeSet<String> },
+    Inactive { removed_in: NonZeroU8 },
+}
+
+impl CurrentVersion {
+    fn version(&self) -> NonZeroU8 {
+        match self {
+            CurrentVersion::Active { version, .. } => *version,
+            CurrentVersion::Inactive { removed_in } => *removed_in,
+        }
+    }
+
+    fn fields(&self) -> Option<&BTreeSet<String>> {
+        match self {
+            CurrentVersion::Active { fields, .. } => Some(fields),
+            CurrentVersion::Inactive { .. } => None,
+        }
+    }
+}
+
+fn defn_error<T>(s: String) -> Result<T, MetricsError> {
```

Comment:
Tiny nit - maybe I'm Stockholm-syndromed into long errors, but I'd probably not have written this function at all, and inlined all the error returns. If you really like this, it could maybe take `s: Into<String>` to save even more call site boilerplate (in the cases where we pass a static string and not a `format!()`'d one).

Reply:
I'm definitely not wedded to this. It was mostly evolution rather than design anyway! I can change it up.

Reply:
Removed in a410171f4